### ISSUE TITLE
Wrong rendering when optimizating

### ISSUE THREAD

**Issue by AndresCasado** [facebookresearch/pytorch3d#236]
## 🐛 Bugs / Unexpected behaviors

I'm trying to combine the example of textured mesh rendering and camera position optimization. The optimization process itself works, but the resulting image is not correct.

Rendering during the optimization process makes the wireframe of the model visible with blurred white lines.

This is the target image that is rendered outside of the optimization loop:
![Imgur Image](http://i.imgur.com/5MmRvMM.jpg)

Then, after optimizing the camera position, the renderer returns this image:
![Imgur Image](http://i.imgur.com/Gs3NR1I.jpg)

## Instructions To Reproduce the Issue:
<details>

<summary>Code</summary>

This is the code used

```python
import matplotlib.pyplot as plt
import numpy as np
import pytorch3d.io as torch3d_io
import pytorch3d.renderer as torch3d_render
import torch
import torch.nn as nn

device = torch.device("cuda")

obj_filename = "cow_mesh/box_subdiv.obj"

mesh = torch3d_io.load_objs_as_meshes([obj_filename], device=device)

R, T = torch3d_render.look_at_view_transform(dist=3.0, elev=35.0, azim=45.0)
cameras = torch3d_render.OpenGLPerspectiveCameras(device=device, R=R, T=T)

raster_settings = torch3d_render.RasterizationSettings(
    image_size=512,
    blur_radius=0.0,
    faces_per_pixel=1,
)

lights = torch3d_render.PointLights(device=device, location=[[0, 5, 0]])

renderer = torch3d_render.MeshRenderer(
    rasterizer=torch3d_render.MeshRasterizer(
        cameras=cameras,
        raster_settings=raster_settings,
    ),
    shader=torch3d_render.TexturedSoftPhongShader(
        device=device,
        cameras=cameras,
        lights=lights,
    )
)

image_reference = renderer(mesh)
plt.figure(figsize=(10, 10,))
plt.imshow(image_reference[0, ..., :3].cpu().numpy())
plt.grid('off')
plt.axis('off')
plt.show()


class Model(nn.Module):
    def __init__(self, meshes, renderer, image_ref):
        super().__init__()
        self.meshes = meshes
        self.device = meshes.device
        self.renderer = renderer

        self.register_buffer('image_ref', image_ref)

        self.camera_position = nn.Parameter(
            torch.from_numpy(np.array([4.1, 4.1, 2.1], dtype=np.float32)).to(self.device)
        )

    def forward(self):
        _R = torch3d_render.look_at_rotation(
            self.camera_position[None, :],
            device=self.device,
        )
        _T = -torch.bmm(
            _R.transpose(1, 2),
            self.camera_position[None, :, None],
        )[:, :, 0]

        image = self.renderer(self.meshes.clone(), R=_R, T=_T)

        loss = torch.sum((image[:3] - self.image_ref[:3]) ** 2)
        return loss, image


model = Model(meshes=mesh, renderer=renderer, image_ref=image_reference)

optimizer = torch.optim.Adam(model.parameters(), lr=0.025)

for i in range(100000):
    optimizer.zero_grad()
    loss, image = model()
    loss.backward()
    optimizer.step()

    print(loss.item())
    print(model.camera_position)
    print()

    n = 200
    if i % n == 0:
        plt.figure(figsize=(10, 10))
        plt.imshow(image.squeeze().cpu().detach().numpy())
        plt.show()

```
</details>

I can provide the model if necessary, but the cow from the examples also shows the same problem.

## System

OS: Ubuntu 20.04
Python 3.8
PyTorch 1.5.0
Torchvision 0.6.0
PyTorch3D compiled from master branch (https://github.com/facebookresearch/pytorch3d/commit/7f1e63aed1252ba8145d4a66ce2272331d60cdae) using gcc 8.4.0, nvcc 10.1.243

**Comment by AndresCasado** [comment]
I've been trying settings on my own until I found the cause is I had not set any `BlendParams`. Setting one with lower than default values for `sigma` and `gamma` makes the white lines smaller, because they're not white lines drawn over the model, they're softened borders of its triangles.

Lowering `sigma` makes the border less soft so it more or less fixes the problem.

I guess the issue can be closed now, but I'd like to ask, what is the correct approach to this? The render still has soft borders. Do I just live with it?

**Comment by nikhilaravi** [comment]
In the `raster_settings` you need to set `blur_radius` > 0.0 and also need to set `faces_per_pixel` > 1. Even though you are using the `TexturedSoftPhongShader` there is no blending happening as you are only using the closest face with no blurring in the xy direction. Have a look at the [camera position optimization tutorial](https://pytorch3d.org/tutorials/camera_position_optimization_with_differentiable_rendering) for how to initialize the `raster_settings`.

**Comment by AndresCasado** [comment]
Thank you for the explanation.

You're right, I was using the settings from the rendering tutorial instead of the optimization tutorial. I figured out that the blurring had something to do with the differentiable capabilities but I didn't know about the rasterization settings.

I got closer to my objective now, but I'm still having some problems. Is there a better place to ask questions? The PyTorch forums? I think my problems are not caused by PyTorch3D (as I thought with this issue) but my lacking knowledge.
